### PR TITLE
MESOS: pass node capacity values to kubelet

### PR DESCRIPTION
**Pull request by sttts**
The node capacity is derived from the Mesos slave resources and passed to the kubelet via our modified cAdvisor.

This patch is based on https://github.com/kubernetes/kubernetes/pull/13036. Only the last few commits are relevant for this PR.

This is only a work-around until https://github.com/kubernetes/kubernetes/issues/13984 hits master.

### CODE REVIEW

File: contrib/mesos/pkg/executor/executor.go [hunk 1]
```diff
@@ -208,23 +207,81 @@ func (k *KubernetesExecutor) isDone() bool {
...
+			case "cpus":
+				k.executorCPU = r.GetScalar().GetValue()
+			case "mem":
+				k.executorMem = r.GetScalar().GetValue()
+			}
+		}
+	}
+
+	// get resource capacity of the node
+	for _, r := range si.GetResources() {
```
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: do we need to check for role="something" here? probably not... but not 100% sure

File: contrib/mesos/pkg/executor/executor.go [hunk 2]
```diff
@@ -208,23 +207,81 @@ func (k *KubernetesExecutor) isDone() bool {
...
+	return
+}
+
+func (k *KubernetesExecutor) onRegister(ei *mesos.ExecutorInfo, si *mesos.SlaveInfo) {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	if !(&k.state).transition(disconnectedState, connectedState) {
+		log.Errorf("failed transitioning to a connected state")
+	}
```
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: i thought "why don't we return here if the state doesn't match?". turns out that the underlying executor driver doesn't ever invoke Disconnected for any reason -- it makes no effort to communicate disconnection events back to the Executor. in apache/mesos, there's a libprocess `exited` func that's invoked when persistent socket connections are severed, mesos-go does no such thing. probably worth noting here, to some degree.

Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: /cc @spacejam

File: contrib/mesos/pkg/executor/executor.go (hunk 1 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: Very good question. Will check.

File: contrib/mesos/pkg/executor/service/service.go [hunk 3]
```diff
@@ -178,7 +181,82 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
...
+	}
+	driver, err := bindings.NewMesosExecutorDriver(dconfig)
+	if err != nil {
+		log.Fatalf("failed to create executor driver: %v", err)
+	}
+	log.V(2).Infof("Initialize executor driver...")
+	exec.Init(driver)
+
+	// start the driver
+	go func() {
```
Comment (jdef) [inline contrib/mesos/pkg/executor/service/service.go]: probably useful to write:
```go
driverStopped := runtime.After(func() { ... }) // via contrib/mesos/pkg/runtime
```
and then, later:
```go
select {
case <-driverStopped:
  // driver aborted before registration complete?! should probably die here.. by returning an error
case <-exec.InitialRegComplete():
  // proceed with initialization
}
```

File: contrib/mesos/pkg/executor/service/service.go [hunk 4]
```diff
@@ -178,7 +181,82 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
...
+		if _, err := driver.Run(); err != nil {
+			log.Fatalf("executor driver failed: %v", err)
```
Comment (jdef) [inline contrib/mesos/pkg/executor/service/service.go]: just `log.Errorf` here instead

**Comment by sttts** [comment]
@jdef this one adds 3 more commits to https://github.com/kubernetes/kubernetes/pull/13036

**Comment by jdef** [comment]
left some comments on the commits

**Comment by sttts** [comment]
@jdef addressed your review comments and rebased.

**Comment by sttts** [comment]
@jdef this one is also green. As before only the last 2 commits are added on-top of the decouple patch. PTAL

**Comment by jdef** [comment]
minor comments, otherwise lgtm

**Comment by sttts** [comment]
Rebased.

**Comment by sttts** [comment]
Rebased onto latest version of https://github.com/kubernetes/kubernetes/pull/13036

**Comment by sttts** [comment]
Rebased after https://github.com/kubernetes/kubernetes/pull/13036 has been merged.

**Pull request merged**